### PR TITLE
Add return type annotations to MultiIndex

### PR DESCRIPTION
**Pull request by mroeschke**
## Description
Mostly just return type annotations. No logic changes.

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/core/multiindex.py [hunk 1]
```diff
@@ -256,7 +256,7 @@ def astype(self, dtype, copy: bool = True):
         return self
 
     @_performance_tracking
-    def rename(self, names, inplace=False):
+    def rename(self, names, inplace: bool = False) -> Self | None:
```
Comment (Matt711) [inline python/cudf/cudf/core/multiindex.py:259]: 1. Does Pandas have its own version of `Self` that it uses for doing type annotations?
2. Do you actively look for where we're missing type annotations, or it more a "you fix as you see it" thing?

Comment (mroeschke) [reply python/cudf/cudf/core/multiindex.py:259]: 1. pandas uses the same `Self` from the `typing_extensions` package https://github.com/pandas-dev/pandas/blob/main/pandas/_typing.py#L102-L107
2. More of the latter, "fix as I see it". This PR is somewhat of an exception. Generally if I modify any part of cudf Python I also try to add "easy" typing annotations around the modification.

Comment (Matt711) [reply python/cudf/cudf/core/multiindex.py:259]: Ah okay, thanks!

**Review by Matt711 — APPROVED** [review]
These changes so far LGTM. I have one question about which arguments we should type annotate. For example, should we type annotate `other` or `sort` here?
```
def union(self, other, sort=None) -> Self:
```

**Comment by mroeschke** [comment]
Yeah ideally all arguments and return types should ideally be annotated (even the obvious `self: Self` should be annotated eventually). It will take a lot of time, but it would be nice eventually to enable some mypy rules where all signatures should be fully annotated

**Pull request merged**